software/contrib/cvecorder.py

```diff
@@ -0,0 +1,348 @@
+
+        @din.handler_falling
+        def endClock():
+            self.handleClock()
+
+        @b1.handler
+        def b1Pressed():
+            # Set recording boolean to true and clear the recording buffer
+            self.CvRecording[self.ActiveCvr] = 'pending'
+            # Clear the array
+            for n in range (0, self.stepLength):
+                self.CVR[self.ActiveBank][self.ActiveCvr][n] = 0
+
+        # B2 Long press
+        @b2.handler_falling
+        def b2PressedLong():
+            # 2000ms press clears all banks
+            if ticks_diff(ticks_ms(), b2.last_pressed()) >  2000:
+                self.clearCvrs('all')
+                #self.bankToSave = self.ActiveBank
+                #self.saveState()
+                # reverse the ActiveCvr increment caused by the initial button press
+                if self.ActiveCvr > 0:
+                    self.ActiveCvr -= 1
+                else:
+                    self.ActiveCvr = self.numCVR
+            # 500ms press clears the active bank
+            elif ticks_diff(ticks_ms(), b2.last_pressed()) >  500:
+                self.clearCvrs(self.ActiveBank)
+                self.bankToSave = self.ActiveBank
+                self.saveState()
+                # reverse the ActiveCvr increment caused by the initial button press
+                if self.ActiveCvr > 0:
+                    self.ActiveCvr -= 1
+                else:
+                    self.ActiveCvr = self.numCVR
+
+        # B2 short press
+        @b2.handler
+        def b2Pressed():
+            # Change the active recorder channel
+            if self.ActiveCvr < self.numCVR:
+                self.ActiveCvr += 1
+            else:
+                self.ActiveCvr = 0
+
+    def handleClock(self):
+
+        # Sample input to 2 decimal places
+        self.CvIn = round(20 * ain.percent(), 2)
+
+        # Start recording if pending and on first step
+        if self.step == 0 and self.CvRecording[self.ActiveCvr] == 'pending':
+            self.CvRecording[self.ActiveCvr] = 'true'
+
+        for i in range(self.numCVR+1):
+            # If recording, write the sampled value to the CVR list and play the voltage
+            if self.CvRecording[i] == 'true':
+                self.CVR[self.ActiveBank][self.ActiveCvr][self.step] = self.CvIn
+                cvs[self.ActiveCvr].voltage(self.CvIn)
+            else:
+                cvs[i].voltage(self.CVR[self.ActiveBank][i][self.step])
+
+        # Reset step number at stepLength -1 as pattern arrays are zero-based
+        if self.step < self.stepLength - 1:
+            self.step += 1
+        else:
+            # Reset step to zero , stop recording and save recording to local storage
+            self.step = 0
+            if self.CvRecording[self.ActiveCvr] == 'true':
+                self.CvRecording[self.ActiveCvr] = 'false'
+                self.bankToSave = self.ActiveBank
+                self.saveState()
+
+    def clearCvrs(self, bank):
```

Comment:
It unnecessarily complicates the code to have the parameter `bank` be either a string ('all') or an int representing the bank number. Consider instead having a parameter `banks` that was a List containing the banks to clear. In most cases it would be a single item, the bank to clear. In the 'all' case, you would generate the list of all banks (using `range`). But in both cases the code would be the same, just iterate through the list and perform the clear. No more special cases!

Reply:
I like this idea, let me have a play with this - thanks!